shared/src/commonMain/kotlin/io/customer/shared/database/QueryHelper.kt

```diff
@@ -0,0 +1,264 @@
+package io.customer.shared.database
+
+import com.squareup.sqldelight.TransactionWithReturn
+import io.customer.shared.common.QueueTaskResult
+import io.customer.shared.common.Success
+import io.customer.shared.runSuspended
+import io.customer.shared.sdk.config.BackgroundQueueConfig
+import io.customer.shared.sdk.meta.Workspace
+import io.customer.shared.tracking.constant.Priority
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.tracking.model.*
+import io.customer.shared.tracking.queue.TaskResultListener
+import io.customer.shared.tracking.queue.failure
+import io.customer.shared.tracking.queue.success
+import io.customer.shared.util.*
+import io.customer.shared.work.BackgroundDispatcher
+import io.customer.shared.work.MainDispatcher
+import kotlinx.datetime.Instant
+import local.TrackingTask
+import local.TrackingTaskQueries
+
+/**
+ * The class works as a bridge for SQL queries. All queries to database should be made using this
+ * class to provide abstraction and keep communication with database easier.
+ */
+internal interface QueryHelper {
+    @MainDispatcher
+    fun insertTask(
+        task: Task,
+        listener: TaskResultListener<QueueTaskResult>? = null,
+    )
+
+    @BackgroundDispatcher
+    fun updateTasksStatus(
+        status: QueueTaskStatus,
+        tasks: List<TrackingTask>,
+    ): QueueTaskResult
+
+    @BackgroundDispatcher
+    fun updateTasksStatusFromResponse(
+        responses: List<TaskResponse>,
+    ): QueueTaskResult
+
+    @BackgroundDispatcher
+    fun selectAllPendingTasks(): List<TrackingTask>?
+
+    @MainDispatcher
```

Comment:
I created `MainDispatcher` and `BackgroundDispatcher` annotations to set the expectations in convenient way (much like the idea of `MainThread` and `WorkerThread`). I initially planned to remove them later but I now see they add some value in quick overview. Do you think we should keep them? Or is there any better solution you can suggest for similar results?

Reply:
The problem is that annotations are easy to ignore while writing code. I prefer to design the API of code to suggest using it in a particular way. 

So, I lean towards saying we don't need the annotations and by having them we are adding more code to maintain in the future. 

Unfortunately, I am unsure until I see this code being called.

Reply:
Agreed. I tried to design them in synchronous and asynchronous way. So any method expected to work in the background, returns the result instantly. While the methods that can run on main dispatchers, always take listener for result.

If we continue following similar conventions in our APIs, I don't see these annotations adding much value. I'm removing them for now. I don't think we would need them in the future too, but if we do, we can reconsider having something similar.